docs: add a design doc proposal for mon zonal availability support (#11407)
rook/rook: design/ceph/ceph-mon-pv.md

knowledge of the scheduling policy and algorithm. The scenario to watch out for
is an endless loop in which the health check is determining a bad placement but
the k8s schedule thinks otherwise.

## Monitor Storage Class Support

It is generally desirable in production to have multiple zones availability. A
whole storage class could go unavailable then and enough mons could stay up to
keep the cluster running.
In such a case the distribution of mons can be made more homogeneous for each
zone, by this we can leverage the spread of mons for providing high
availability. This could be a useful scenario for a datacenter with own storage
class in contrast to ones provided by cloud providers.

> Rook will enable the mon deployment when requested by the admin through the

Reviewer: The section below seems like it already shows how the CRD works, so I think we can remove this sentence.
```suggestion
```

usual Rook CRDs.

To support this scenario, we can do something similar to the mon deployment for
[Stretch clusters](https://github.com/ceph/ceph/blob/master/doc/rados/operations/stretch-mode.rst).
The operator will have zonal awareness, associating mon deployments based on
zones configuration provided.

The type of failure domain used for this scenario is commonly "zone", but can
be set to a different failure domain.

### Failure Domain Configuration

The topology of the K8s cluster is to be determined by the admin, outside the scope of Rook.
Rook will simply detect the topology labels that have been added to the nodes.

If the desired failure domain is a "zone", the `topology.kubernetes.io/zone` label should
be added to the nodes. Any of the [topology labels](https://rook.io/docs/rook/latest/ceph-cluster-crd.html#osd-topology)

> supported by OSDs can be used.

Reviewer: ```suggestion
supported by OSDs can be used also for this scenario.
```

For example:

```yaml
topology.kubernetes.io/zone=a
topology.kubernetes.io/zone=b
topology.kubernetes.io/zone=c
```

###  Design

The core changes to rook are to associate the mons with the required zones.

> - Each zone will have at least one mon assigned using labels for nodeAffinity.

Reviewer: ```suggestion
- Each zone will have at most one mon assigned using labels for nodeAffinity. If there are more zones than mons, some zones may not have a mon. 
```

The new configuration will support the `zones` configuration where the
failure domains available must be listed.

```yaml
  mon:
    count: 3
    allowMultiplePerNode: false
    zones:
      - name: us-east-2a
      - name: us-east-2b
      - name: us-east-2c

>     volumeClaimTemplate:

Reviewer: The scenario really requires a volumeClaimTemplate to be specified for each zone. So how about we replace this example with the one later in the doc. One example seems sufficient.

      spec:
        storageClassName: zone-a-storage
        resources:
          requests:
            storage: 10Gi
```

> The operator will track which mons belong to which zone. The zone assignment for each mon

Reviewer: We don't need to track the zone assignment for each mon anymore. That's just for stretch clusters. So we can remove this section about the configmap.

will be stored in the `rook-ceph-mon-endpoints` configmap in the same data structure where the host
assignment is stored for mons with node affinity.

For example, if the zones are called
"zone-a", "zone-b", and "zone-c", the configmap would contain:

```
data:
  data: a=10.99.109.200:6789,b=10.98.18.147:6789,c=10.96.86.248:6789,d=10.96.86.249:6789,e=10.96.86.250:6789
  mapping: '{"node":{"a":"zone-a","b":"zone-b","c":"zone-c"}}'
  maxMonId: "3"
```

If no zones are listed in the `spec.mon.zones`, It will default to arbitrary
deployment without zonal awareness.

> ### Mon Failover

Reviewer: The paragraph about the mon canary just seems like part of the main design, no need for the failover heading.
```suggestion
```

If the mon.zones are specified, the mon canary pod should not specify the volume
for the mon store. The whole purpose of the mon canaries is to allow the k8s
scheduler to decide where the mon should be scheduled.
So if we don't add a volume to the canary pod, we can delay the creation of the
mon PVC until the canary pod is done. Then after the mon canary is done
scheduling, the operator can look at the node where it was scheduled and
determine which zone (or other topology) the node belonged to.
For the mon daemon pod, then the operator would create the PVC from the
volumeClaimTemplate matching the zone where the mon canary was created.
This way mon canary pod will be able to take care of all scheduling tasks.

> ### Mon Storage Configuration

Reviewer: Let's remove this section since the examples will consolidate above.

For example:

```yaml
topology.kubernetes.io/zone=a
topology.kubernetes.io/zone=b
topology.kubernetes.io/zone=c
```

###  Design

The core changes to rook are to associate the mons with the required zones.

- Each zone will have at most one mon assigned using labels for nodeAffinity.
 If there are more zones than mons, some zones may not have a mon.

The new configuration will support the `zones` configuration where the
failure domains available must be listed.

If the mon.zones are specified, the mon canary pod should not specify the volume
for the mon store. The whole purpose of the mon canaries is to allow the k8s
scheduler to decide where the mon should be scheduled.
So if we don't add a volume to the canary pod, we can delay the creation of the
mon PVC until the canary pod is done. Then after the mon canary is done
scheduling, the operator can look at the node where it was scheduled and
determine which zone (or other topology) the node belonged to.
For the mon daemon pod, then the operator would create the PVC from the
volumeClaimTemplate matching the zone where the mon canary was created.
This way mon canary pod will be able to take care of all scheduling tasks.

> In this example, all the mons will be backed by default volumeClaimTemplate:

Reviewer: ?

> In this example, all the mons will be backed by default volumeClaimTemplate:

Author: aah, missed, updated thanks!
